### DISCUSSION TITLE
Multi-threaded testing with separate DB connections and teardown of before_all

### DISCUSSION THREAD

**Discussion by cou929** [test-prof/test-prof#232] [Q&A]
Suppose that I have a test cases that uses Thread class with multiple DB connections, and another test case that uses test-prof's before_all. When I run these, the former test does not success.

For example, let's say I have a Foo model that uses MySQL's GET_LOCK to update data like this:

```ruby
class Foo < ApplicationRecord
  def update_carefully
    lock_str = "FOO_LOCK_STR"

    ActiveRecord::Base.connected_to(role: :writing) do # get lock to primary instance
      if ActiveRecord::Base.connection.get_advisory_lock(lock_str, 5)
        self.reload
        raise "#{self.id} is already updated" if self.col1 == "updated"

        self.update!(col1: "updated")
      else
        raise "Failed to get lock"
      end
    end
  ensure
    ActiveRecord::Base.connected_to(role: :writing) do
      ActiveRecord::Base.connection.release_advisory_lock(lock_str)
    end
  end
end
```

To test this Foo model, we will use the Thread class to call `update_carefully` from several separate DB connections and make sure that an exception is raised.

```ruby
# spec/foo.rb
require "rails_helper"

describe "thread test" do
  let (:foo) { FactoryBot.create(:foo) }
  it "takes lock before updating" do
    expect {
      threads = 2.times.map do
        Thread.new do
          ActiveRecord::Base.connection_pool.with_connection do
            foo.update_carefully
          end
        end
      end
      threads.each(&:join)
    }.to raise_error
  end
end
```

Apart from this, suppose I have an existing test case that uses before_all.

```ruby
# spec/other.rb
require "rails_helper"

RSpec.describe "test" do
  before_all do
    p "before_all"
  end

  it "test" do
    expect(1).to eq(1)
  end
end
```

When I run these two tests, `spec/foo.rb` sometimes fails.

```sh
$ rspec spec/other.rb spec/foo.rb
...
# Expected errors in spec/foo.rb do not raise
```

The reason is that in spec/other.rb before_all, [test-prof forces to use the DB connection from the main thread](https://github.com/test-prof/test-prof/blob/bcddd55bb6c4f31da0e473e6177a73d3b3d1d37c/lib/test_prof/before_all/adapters/active_record.rb#L47) by [lock_thread method](https://github.com/rails/rails/blob/d3b602b9093d81d1cf92e11cbbe879a5d035f179/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L167-L173), and the child threads created by Thread.new use the same connection as the main thread, so the intended test cannot be performed. There is no teardown such like lock_thread = false, and the lock_thread continues in another test case, causing spec/foo.rb to fail. In fact, in this example, if I only run spec/foo.rb, the test will pass.

It is possible to work around this by clearing the lock_thread in spec/foo.rb.

```ruby
# spec/foo.rb
require "rails_helper"

describe "thread test" do
  let (:foo) { FactoryBot.create(:foo) }

  # add these
  before do
    ActiveRecord::Base.connection.pool.lock_thread = false
  end

  it "takes lock before updating" do
    # ...
  end
end
```

So my question is the following.

- Is this behavior (lock_thread changes are continued even in test cases that do not use test-prof) what test-prof is intended to do?
- Is there a more appropriate way to deal with this than the above workarounds? (For example, does test-prof have an option to control lock_threads?)

**Comment by palkan** [comment]
> Is this behavior (lock_thread changes are continued even in test cases that do not use test-prof) what test-prof is intended to do?

Nope. I haven't thought about it. However, `before_all` is assumed to be use with transaction tests, which means, we expect Rails to set `lock_thread=true` (the only reason why we do that manually is that Rails does this later than we need).

Do you use Rails `transactional_tests = true`? If so, your test would be affected by `lock_thread` even without `before_all`.

I suggest making `lock_thread=false` explicit in the test itself (almost like you did); maybe, with a helper like this:

```rb
def without_db_thread_lock
  was_locked = ActiveRecord::Base.connection.pool.lock_thread
  ActiveRecord::Base.connection.pool.lock_thread = false
  yield
ensure
  ActiveRecord::Base.connection.pool.lock_thread = was_locked
end

# and then
without_db_thread_lock do
  threads = 2.times.map do
    Thread.new do
       ActiveRecord::Base.connection_pool.with_connection do
         foo.update_carefully
       end
     end
  end
end
```

**Reply by cou929** [reply]
> Do you use Rails transactional_tests = true?

No, in my case I always set `transactional_tests = false` and use [database_cleaner](https://github.com/DatabaseCleaner/database_cleaner) to manage the transactional tests. Basically, I always run transactional tests with database_cleaner (`DatabaseCleaner.strategy = :transaction`) and disable it only in some cases (setting `strategy = :truncation`). For example, the aforementioned `spec/other.rb` runs with `strategy = :truncation` and `spec/foo.rb` runs with `strategy = :transaction`.

My expectation is that test-prof will only set `lock_thread = true` while it's being used, and in other cases it will run with the default value of `lock_thread`. test-prof currently set `lock_thread = true` in the before hook of before_all, but it would be more intuitive and easier to understand for me if `lock_thread` were set back to its original value in the equivalent of the after hook, for example. Is this makes sense?

By the way, your suggestion of `without_db_thread_lock` is very useful and I will use it. Thank you :)

**Reply by cou929** [reply]
@palkan What is the background of not returning the original value to ActiveRecord's lock_thread in the `after` (teardown) equivalent sequence?

Is it reasonable to use `after` to restore the original value like the following?

```ruby
# lib/test_prof/before_all/adapters/active_record.rb

config.after(:begin) do
    next unless ::ActiveRecord::Base.connection.pool.respond_to?(:lock_thread=)
    ::ActiveRecord::Base.connection.pool.lock_thread = orig_value
end
```

In ActiveRecord, the lock_thread is [restored during teardown](https://github.com/rails/rails/blob/2b3af27d8232db8def0c6600e89a714287381e18/activerecord/lib/active_record/test_fixtures.rb#L171). The Rails guide [says](https://guides.rubyonrails.org/testing.html#testing-parallel-transactions) to set `use_transactional_tests = false` when testing parallel transactions. For this purpose, lock_thread must be restored after use in this way. The same behavior in test-prof seems more natural to me.

If this is makes sense, I can file an issue or create a pr. What do you think?

**Reply by palkan** [reply]
>  lock_thread must be restored after use in this way.

yeah, makes sense; a PR would be great.

> config.after(:begin) do

I think, it should be `config.after(:rollback)`.
